adapters/rack/lib/opentelemetry/adapters/rack/middlewares/tracer_middleware.rb

```diff
@@ -0,0 +1,257 @@
+                                                                                   full_path: full_path(rack_request),
+                                                                                   base_url: base_url(rack_request)),
+                                               kind: :server)
+
+              begin
+                @app.call(env).tap do |status, headers, response|
+                  set_attributes_after_request(request_span, status, headers, response)
+                end
+              rescue StandardError => e
+                record_and_reraise_error(e, request_span: request_span)
+              ensure
+                request_span.finish
+                frontend_span&.finish
+              end
+            end
+          end
+
+          private
+
+          # return Context
+          def create_frontend_span(env)
+            # NOTE: get_request_start may return nil
+            request_start_time = OpenTelemetry::Adapters::Rack::Util::QueueTime.get_request_start(env)
+            record_frontend_span = config[:record_frontend_span] && !request_start_time.nil?
+
+            return unless record_frontend_span
+
+            span = tracer.start_span('http_server.queue', # NOTE: span kind of 'proxy' is not defined
+                                     # NOTE: initialize with as many attributes as possible:
+                                     attributes: {
+                                       'component' => 'http',
+                                       'service' => config[:web_service_name],
+                                       'start_time' => request_start_time
+                                     },
+                                     kind: :server)
+
+            span
+          end
+
+          def current_span_key
+            OpenTelemetry::Trace::Propagation::ContextKeys.current_span_key
+          end
+
+          def tracer
+            OpenTelemetry::Adapters::Rack::Adapter.instance.tracer
+          end
+
+          ### request_span
+
+          # Per https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-http.md#http-server-semantic-conventions
+          #
+          # One of the following sets is required, in descending preference:
+          # * http.scheme, http.host, http.target
+          # * http.scheme, http.server_name, net.host.port, http.target
+          # * http.scheme, net.host.name, net.host.port, http.target
+          # * http.url
+          def request_span_attributes(env:, full_http_request_url:, full_path:, base_url:)
+            {
+              'component' => 'http',
+              'http.method' => env['REQUEST_METHOD'],
+              'http.url' => full_http_request_url,
+              'http.host' => env['HOST'],
+              'http.scheme' => env['rack.url_scheme'],
+              'http.target' => full_path,
+              'http.base_url' => base_url # NOTE: 'http.base_url' isn't officially defined
```

Comment:
I think we can just go with `http.scheme`, `http.host` and `http.target` as per the semantic conventions.

Reply:
I've removed the extra `http.url` and `http.base_url` attributes.